OpenRA.Mods.Common/Traits/World/ResourceLayer.cs

```diff
@@ -10,204 +10,308 @@
-					var temp = Content[cell];
-					temp.Density = Math.Max(density, 1);
-
-					Content[cell] = temp;
+					var c = cell + directions[i];
+					if (Content.Contains(c) && Content[c].Type == resource.Type)
+						++adjacent;
 				}
+
+				// Adjacent includes the current cell, so is always >= 1
+				var density = Math.Max(int2.Lerp(0, resourceInfo.MaxDensity, adjacent, 9), 1);
+				Content[cell] = new ResourceLayerContents(resource.Type, density);
 			}
 		}
 
-		public bool AllowResourceAt(ResourceType rt, CPos cell)
+		void IWorldLoaded.WorldLoaded(World w, WorldRenderer wr) { WorldLoaded(w, wr); }
+
+		protected virtual bool AllowResourceAt(string resourceType, CPos cell)
 		{
-			if (!world.Map.Contains(cell))
+			if (!Map.Contains(cell) || Map.Ramp[cell] != 0)
 				return false;
 
-			if (!rt.Info.AllowedTerrainTypes.Contains(world.Map.GetTerrainInfo(cell).Type))
+			if (resourceType == null || !info.ResourceTypes.TryGetValue(resourceType, out var resourceInfo))
 				return false;
 
-			if (!rt.Info.AllowUnderActors && world.ActorMap.AnyActorsAt(cell))
+			if (!resourceInfo.AllowedTerrainTypes.Contains(Map.GetTerrainInfo(cell).Type))
 				return false;
 
-			if (!rt.Info.AllowUnderBuildings && buildingInfluence.GetBuildingAt(cell) != null)
-				return false;
+			return BuildingInfluence.GetBuildingAt(cell) == null;
+		}
+
+		ResourceLayerContents CreateResourceCell(string resourceType, CPos cell, int density)
+		{
+			if (!info.ResourceTypes.TryGetValue(resourceType, out var resourceInfo))
+			{
+				world.Map.CustomTerrain[cell] = byte.MaxValue;
+				return ResourceLayerContents.Empty;
+			}
 
-			return rt.Info.AllowOnRamps || world.Map.Ramp[cell] == 0;
+			world.Map.CustomTerrain[cell] = world.Map.Rules.TerrainInfo.GetTerrainIndex(resourceInfo.TerrainType);
+			++resCells;
+
+			return new ResourceLayerContents(resourceType, density.Clamp(1, resourceInfo.MaxDensity));
 		}
 
-		public bool CanSpawnResourceAt(ResourceType newResourceType, CPos cell)
+		bool CanAddResource(string resourceType, CPos cell, int amount = 1)
 		{
 			if (!world.Map.Contains(cell))
 				return false;
 
-			var currentResourceType = GetResourceType(cell);
-			return (currentResourceType == newResourceType && !IsFull(cell))
-				|| (currentResourceType == null && AllowResourceAt(newResourceType, cell));
-		}
+			if (resourceType == null || !info.ResourceTypes.TryGetValue(resourceType, out var resourceInfo))
+				return false;
 
-		ResourceLayerContents CreateResourceCell(ResourceType t, CPos cell)
-		{
-			world.Map.CustomTerrain[cell] = world.Map.Rules.TerrainInfo.GetTerrainIndex(t.Info.TerrainType);
-			++resCells;
+			var content = Content[cell];
+			if (content.Type == null)
+				return amount <= resourceInfo.MaxDensity && AllowResourceAt(resourceType, cell);
 
-			return new ResourceLayerContents
-			{
-				Type = t
-			};
+			if (content.Type != resourceType)
+				return false;
+
+			return content.Density + amount <= resourceInfo.MaxDensity;
```

Comment:
`AddResource` below would add regardles by capping at `MaxDensity`. This should probably be
```suggestion
			return content.Density <= resourceInfo.MaxDensity;
```

Reply:
I agree in theory, but in practice this would break resource spawning. I'm doing my best in this PR to work around badly designed plumbing, without having to burn everything to the ground and restart from scratch.